bundles/auth/org.eclipse.smarthome.auth.oauth2client/src/main/java/org/eclipse/smarthome/auth/oauth2client/internal/OAuthStoreHandlerImpl.java

```diff
@@ -0,0 +1,531 @@
+import org.eclipse.smarthome.core.auth.client.oauth2.AccessTokenResponse;
+import org.eclipse.smarthome.core.auth.client.oauth2.StorageCipher;
+import org.eclipse.smarthome.core.storage.Storage;
+import org.eclipse.smarthome.core.storage.StorageService;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.RemovalListener;
+import com.google.common.cache.RemovalNotification;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonDeserializer;
+import com.google.gson.JsonPrimitive;
+import com.google.gson.JsonSerializer;
+
+/**
+ * This class handles the storage directly. It is internal to the OAuthClientService and there is
+ * little need to study this.
+ *
+ * The first role of this handler storing and caching the access token response, and persisted parameters.
+ *
+ * The storage contains these:
+ * 1. INDEX_HANDLES = json string-set of all handles
+ * 2. <handle>.LastUsed = system-time-milliseconds
+ * 3. <handle>.AccessTokenResponse = Json of AccessTokenResponse
+ * 4. <handle>.ServiceConfiguration = Json of PersistedParameters
+ *
+ * If at any time, the storage is not available, it is still possible to read existing access tokens from store.
+ * The last-used statistics for this access token is broken. It is a measured risk to take.
+ *
+ * If at any time, the storage is not available, it is not able to write any new access tokens into store.
+ *
+ * All entries are subject to removal if they have not been used for 183 days or more (half year).
+ * The recycle is performed when then instance is deactivated
+ *
+ * @author Gary Tse - Initial Contribution
+ *
+ */
+@Component(service = OAuthStoreHandler.class, property = {
+        "CIPHER_TARGET=org.eclipse.smarthome.auth.oauth2client.internal.cipher.SymmetricKeyCipher" })
+public class OAuthStoreHandlerImpl implements OAuthStoreHandler {
+
+    // easy mocking with protected access
+    protected static final int EXPIRE_DAYS = 183;
+    protected static final int ACCESS_TOKEN_CACHE_SIZE = 50;
+
+    private static final String STORE_NAME = "StorageHandler.For.OAuthClientService";
+
+    @NonNull
+    private static final String STORE_KEY_INDEX_OF_HANDLES = "INDEX_HANDLES";
+
+    private static final String EXCEPTION_STORAGE = "Unable to continue, storage is not available";
+
+    private StorageService _storageServiceDoNotUseDirectly;
+    private Storage<String> _storeDoNotUseDirectly;
```

Comment:
I am not sure if the variable naming convention in ESH includes the use of underscore. It is definitely used for constants but I am note sure if it is really followed in ESH. You can cross-check with other guys.